src/hotspot/share/opto/intrinsicnode.cpp

```diff
@@ -261,27 +261,37 @@ static const Type* bitshuffle_value(const TypeInteger* src_type, const TypeInteg
   }
 
   if (!mask_type->is_con()) {
-    int mask_max_bw;
-    int max_bw = bt == T_INT ? 32 : 64;
-    // Case 1) Mask value range includes -1.
-    if ((mask_type->lo_as_long() < 0L && mask_type->hi_as_long() >= -1L)) {
-      mask_max_bw = max_bw;
-    // Case 2) Mask value range is less than -1.
-    } else if (mask_type->hi_as_long() < -1L) {
-      mask_max_bw = max_bw - 1;
-    } else {
-    // Case 3) Mask value range only includes +ve values.
-      assert(mask_type->lo_as_long() >= 0, "");
-      jlong clz = count_leading_zeros(mask_type->hi_as_long());
-      clz = bt == T_INT ? clz - 32 : clz;
-      mask_max_bw = max_bw - clz;
-    }
     if ( opc == Op_CompressBits) {
+      int mask_max_bw;
+      int max_bw = bt == T_INT ? 32 : 64;
+      // Case 1) Mask value range includes -1, this negates the possibility of
+      // strictly non-negative result value range.
+      if ((mask_type->lo_as_long() < 0L && mask_type->hi_as_long() >= -1L)) {
+        mask_max_bw = max_bw;
+      // Case 2) Mask value range is less than -1, this indicates presence of at least
+      // one zero bit in the mask value, there by constraining the result of compression to
+      // a +ve value range.
+      } else if (mask_type->hi_as_long() < -1L) {
+        mask_max_bw = max_bw - 1;
+      } else {
+      // Case 3) Mask value range only includes +ve values, this can again be
+      // used to ascertain known Zero bits of resultant value.
+        assert(mask_type->lo_as_long() >= 0, "");
+        jlong clz = count_leading_zeros(mask_type->hi_as_long());
```

Comment:
```suggestion
        jlong clz = count_leading_zeros(mask_type->hi_as_long());
        // The mask has at least clz leading zeros, and hence also the compression
        // result must have at least clz leading zeros.
```

Reply:
I think a comment like this is still missing. You should somehow say that the leading zeros in the mask translate to leading zeros in the result.